### PR TITLE
Adding BalanceBuilderPropertyTest

### LINKED ISSUE #3143: [Domain] Build the `AccountBalanceUseCase`
### Please confirm the following

- [X] I checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.

### What would you like to improve?

- Build the `AccountBalanceUseCase`.
- Use the `AccountStatsUseCase` and the `TransactionRepository`

```kotlin
class AccountBalanceUseCase @Inject constructor(...) {
   fun calculate(account: AccountId): Map<AssetCode, NonZeroDouble>
}
```

- Build the `NonZeroDouble` following the pattern to the `PositiveDouble`
- We neee NonZeroDouble to handle the case of negative balances (credits to @Vishwa-Raghavendra)

### Because...

- Proper domain logic following the architecture and having unit + property-based tests.

### Description

_No response_

### Success Criteria

- Follows all dev guidelines (this is a critical code so expect Requests for changes on your PR)
- Property-based tests
- Unit tests

### PR DESCRIPTION
**Pull request by rodrigoliveirac**
## Pull request (PR) checklist

Please check if your pull request fulfills the following requirements:

- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read and understand the [Developer Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Guidelines.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] My PR includes only the necessary changes to fix the issue (i.e., no unnecessary files or lines of code are changed).
- [x] 🎬 I've attached a **screen recording** of using the new code to the next paragraph (if applicable).

## What's changed?

Describe with a few bullets **what's new:**
- I've added property test cases...

## Does this PR close any GitHub issues? (do not delete)

- Closes #3143 

## Troubleshooting GitHub Actions (CI) failures ❌
Pull request checks failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
+package com.ivy.domain.usecase
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt:7]: Why is this file created again? Didn't we merge it already?

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                processWithdrawals(expenses = expenses.values, transfersOut = transferOut.values)
+                build()
+            }[AssetCode.EUR] shouldBe totalEUR
+        }
+    }
+
+    @Test
+    fun `property - check all USD currency`() = runTest {
+        // given
+        val incomes = Arb.statSummary()
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt:120]: No need for this `val`s

File: shared/domain/src/test/java/com/ivy/domain/ArbStatSummary.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import com.ivy.data.model.testing.value
+import com.ivy.domain.model.StatSummary
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.bind
+import io.kotest.property.arbitrary.map
+import io.kotest.property.arbitrary.nonNegativeInt
+
+fun Arb.Companion.statSummary(): Arb<StatSummary> {
+    return Arb.bind(
+        Arb.nonNegativeInt(),
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/ArbStatSummary.kt:14]: We don't use this, remove it. Also look how I build Arbs in the project - there was a nicer way

File: shared/domain/src/test/java/com/ivy/domain/ArbStatSummary.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        Arb.map(keyArb = assetCode(), valueArb = Arb.value().map { it.amount })
+    ) { _, trns ->
+        val nonNegativeInt =
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/ArbStatSummary.kt:17]: You can directly call `NonNegativeInt.unssfe(trns.size)`

Comment (ILIYANGERMANOV) [reply shared/domain/src/test/java/com/ivy/domain/ArbStatSummary.kt:17]: The size of trns will never be <0

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        val arbIrrelevantTransaction = Arb.transaction().filter { trn ->
+            trn.getFromAccount() != account && trn.getToAccount() != account
+        }
+
+        checkAll(Arb.list(arbIrrelevantTransaction)) { trns ->
+            // when
+            val stats = statsUseCase.calculate(account, trns)
+
+            // then
+            BalanceBuilder().run {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt:48]: It make mor sense to directly test the AccountBalanceUseCase with this

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    fun setup() {
+        statsUseCase = AccountStatsUseCase(
+            dispatchers = TestDispatchersProvider,
+            accountRepository = mockk(),
+            exchangeUseCase = mockk(),
+        )
+    }
+
+    @Test
+    fun `property - ignores irrelevant transactions`() = runTest {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt:36]: this is a good property! 💯

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            BalanceBuilder().run {
+                processDeposits(incomes = stats.income.values, transfersIn = stats.transfersIn.values)
+                processWithdrawals(expenses = stats.expense.values, transfersOut = stats.transfersOut.values)
+                build()
+            } shouldBe emptyMap()
+        }
+    }
+
+    @Test
+    fun `property - check all BGP currency`() = runTest {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt:57]: the below properties are too complex to understand and doesn't bring value

File: shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.filter
+import io.kotest.property.arbitrary.list
+import io.kotest.property.checkAll
+import io.mockk.mockk
+import kotlinx.coroutines.test.runTest
+import org.junit.Before
+import org.junit.Test
+
+class BalanceBuilderPropertyTest {
```
Comment (ILIYANGERMANOV) [inline shared/domain/src/test/java/com/ivy/domain/usecase/balance/BalanceBuilderPropertyTest.kt:22]: Add test for:

**property - order of transactions doesn't matter**
- generate an arbitrary list of transactions (original)
- shuffle them (shuffled)
- veify that the balance of `original` and `shuffled` is the same

Again forget the BalanceBuilder, let's make property-tests on the AccountBalanceUseCase

**property - zero balance when each trn is reverted**
- generate an arb list of transactions (trns)
- for each transaction create a 2nd "reverse transaction" (e.g. Income -> Expense, Transfer -> just swap from-to)
- expect that when you calculate balance of both list that it's zero
- verify that balance(trns) == -balance(reverse)
- balance(trna + reverse) == 0

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
Please check that your PR is rebased properly and that we're not duplicating files

**Comment by rodrigoliveirac** [comment]
> Please check that your PR is rebased properly and that we're not duplicating files

I was worried about that too, but I fixed it last week. The correct commits are here: https://github.com/Ivy-Apps/ivy-wallet/pull/3447/commits. It seems like in our conversation it sounded like I made a duplicate commit

**Pull request closed without merging**